File: examples/go/hand/agg/handlers/award_pot.go
package handlers

import (
	"time"

	angzarr "github.com/benjaminabbitt/angzarr/client/go"
	pb "github.com/benjaminabbitt/angzarr/client/go/proto/angzarr"
	"github.com/benjaminabbitt/angzarr/client/go/proto/examples"
	"google.golang.org/protobuf/proto"
	"google.golang.org/protobuf/types/known/anypb"
	"google.golang.org/protobuf/types/known/timestamppb"
)

func guardAwardPot(state HandState) error {
	if !state.Exists() {
		return angzarr.NewCommandRejectedError("Hand does not exist")
	}
	if state.IsComplete() {
		return angzarr.NewCommandRejectedError("Hand already complete")
	}
	return nil
}

func validateAwardPot(cmd *examples.AwardPot, state HandState) error {
	if len(cmd.Awards) == 0 {
		return angzarr.NewCommandRejectedError("No awards specified")
	}

	totalAwarded := int64(0)
	for _, award := range cmd.Awards {
		player := state.GetPlayerByRoot(award.PlayerRoot)
		if player == nil {
			return angzarr.NewCommandRejectedError("Award to player not in hand")
		}
		if player.HasFolded {
			return angzarr.NewCommandRejectedError("Cannot award to folded player")
		}
		totalAwarded += award.Amount
	}

	if totalAwarded > state.TotalPot() {
		return angzarr.NewCommandRejectedError("Awards exceed pot total")
	}

	return nil
}

func computePotAwarded(cmd *examples.AwardPot, now time.Time) (*examples.PotAwarded, []*examples.PotWinner) {
	winners := make([]*examples.PotWinner, len(cmd.Awards))
	for i, award := range cmd.Awards {
		winners[i] = &examples.PotWinner{
			PlayerRoot: award.PlayerRoot,
			Amount:     award.Amount,
			PotType:    award.PotType,
		}
	}

	return &examples.PotAwarded{
		Winners:   winners,
		AwardedAt: timestamppb.New(now),
	}, winners
}

func computeHandComplete(cmd *examples.AwardPot, state HandState, winners []*examples.PotWinner, now time.Time) *examples.HandComplete {
	finalStacks := make([]*examples.PlayerStackSnapshot, 0, len(state.Players))
	for _, player := range state.Players {
		finalStack := player.Stack
		for _, award := range cmd.Awards {
			if state.GetPlayerByRoot(award.PlayerRoot) == player {
				finalStack += award.Amount
			}
		}
		finalStacks = append(finalStacks, &examples.PlayerStackSnapshot{
			PlayerRoot: player.PlayerRoot,
			Stack:      finalStack,
			IsAllIn:    player.IsAllIn,
			HasFolded:  player.HasFolded,
		})
	}

	return &examples.HandComplete{
		TableRoot:   state.TableRoot,
		HandNumber:  state.HandNumber,
		Winners:     winners,
		FinalStacks: finalStacks,
		CompletedAt: timestamppb.New(now),
	}
}

// HandleAwardPot handles awarding the pot to winners.
func HandleAwardPot(
	commandBook *pb.CommandBook,
	commandAny *anypb.Any,
	state HandState,
	seq uint32,
) (*pb.EventBook, error) {
	var cmd examples.AwardPot
	if err := proto.Unmarshal(commandAny.Value, &cmd); err != nil {
		return nil, err
	}

	if err := guardAwardPot(state); err != nil {
		return nil, err
	}
	if err := validateAwardPot(&cmd, state); err != nil {
		return nil, err
	}

	now := time.Now()
	potAwardedEvent, winners := computePotAwarded(&cmd, now)
	handCompleteEvent := computeHandComplete(&cmd, state, winners, now)

	potAwardedAny, err := anypb.New(potAwardedEvent)
	if err != nil {
		return nil, err
	}

	handCompleteAny, err := anypb.New(handCompleteEvent)
	if err != nil {
		return nil, err
	}

	return angzarr.NewEventBookMulti(commandBook.Cover, seq, potAwardedAny, handCompleteAny), nil
}

File: examples/rust/hand/agg/src/handlers/award_pot.rs
//! AwardPot command handler.

use angzarr_client::proto::examples::{
    AwardPot, HandComplete, PlayerStackSnapshot, PotAwarded, PotWinner,
};
use angzarr_client::proto::{CommandBook, EventBook};
use angzarr_client::{new_event_book_multi, pack_event, CommandRejectedError, CommandResult, UnpackAny};
use prost_types::Any;

use crate::state::HandState;

fn guard(state: &HandState) -> CommandResult<()> {
    if !state.exists() {
        return Err(CommandRejectedError::new("Hand does not exist"));
    }
    if state.is_complete() {
        return Err(CommandRejectedError::new("Hand already complete"));
    }
    Ok(())
}

fn validate(cmd: &AwardPot, state: &HandState) -> CommandResult<()> {
    if cmd.awards.is_empty() {
        return Err(CommandRejectedError::new("No awards specified"));
    }

    let mut total_awarded = 0i64;
    for award in &cmd.awards {
        let player = state
            .get_player(&award.player_root)
            .ok_or_else(|| CommandRejectedError::new("Award to player not in hand"))?;

        if player.has_folded {
            return Err(CommandRejectedError::new("Cannot award to folded player"));
        }
        total_awarded += award.amount;
    }

    if total_awarded > state.total_pot() {
        return Err(CommandRejectedError::new("Awards exceed pot total"));
    }

    Ok(())
}

fn compute(cmd: &AwardPot, state: &HandState) -> (PotAwarded, HandComplete) {
    let winners: Vec<PotWinner> = cmd
        .awards
        .iter()
        .map(|award| PotWinner {
            player_root: award.player_root.clone(),
            amount: award.amount,
            pot_type: award.pot_type.clone(),
            winning_hand: None,
        })
        .collect();

    let now = angzarr_client::now();

    let final_stacks: Vec<PlayerStackSnapshot> = state
        .players
        .values()
        .map(|player| {
            let mut final_stack = player.stack;
            for award in &cmd.awards {
                if award.player_root == player.player_root {
                    final_stack += award.amount;
                }
            }
            PlayerStackSnapshot {
                player_root: player.player_root.clone(),
                stack: final_stack,
                is_all_in: player.is_all_in,
                has_folded: player.has_folded,
            }
        })
        .collect();

    let pot_awarded = PotAwarded {
        winners: winners.clone(),
        awarded_at: Some(now.clone()),
    };

    let hand_complete = HandComplete {
        table_root: state.table_root.clone(),
        hand_number: state.hand_number,
        winners,
        final_stacks,
        completed_at: Some(now),
    };

    (pot_awarded, hand_complete)
}

pub fn handle_award_pot(
    command_book: &CommandBook,
    command_any: &Any,
    state: &HandState,
    seq: u32,
) -> CommandResult<EventBook> {
    let cmd: AwardPot = command_any
        .unpack()
        .map_err(|e| CommandRejectedError::new(format!("Failed to decode command: {}", e)))?;

    guard(state)?;
    validate(&cmd, state)?;

    let (pot_awarded, hand_complete) = compute(&cmd, state);

    Ok(new_event_book_multi(
        command_book,
        seq,
        vec![
            pack_event(&pot_awarded, "examples.PotAwarded"),
            pack_event(&hand_complete, "examples.HandComplete"),
        ],
    ))
}
